### PR TITLE
create `@automattic/js-utils` package and add `uniqueBy`

### PR DESCRIPTION
**Pull request by flootr**
#### Changes proposed in this Pull Request

* Create new `@automattic/js-utils` package
* Add `uniqueBy` utility function

(mostly) extracted from #50077 

#### Testing instructions

* Make sure unit tests pass

### CODE REVIEW

File: client/lib/unique-by/test/unique-by.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		expect( uniqueBy( values ) ).toEqual( [ 'a', 'b', 'c', 'd' ] );
+	} );
+
+	test( 'should remove duplicate values by identity by default (mixed numbers & strings)', () => {
+		const values = [ 1, 2, 'a', 'b', 1, 2, 'c', 'a' ];
+		expect( uniqueBy( values ) ).toEqual( [ 1, 2, 'a', 'b', 'c' ] );
+	} );
+
+	test( 'should be able to handle `null` and `undefined` values', () => {
+		// @TODO: I'm not sure about this 🤔
```
Comment (flootr) [inline client/lib/unique-by/test/unique-by.js]: 🤔

**Comment by flootr** [comment]
More or less copied from [30secondsofcode.org/js/s/unique-elements-by](https://www.30secondsofcode.org/js/s/unique-elements-by)

File: client/lib/unique-by/test/unique-by.js (hunk 1 above)
Comment (flootr) [reply client/lib/unique-by/test/unique-by.js]: Behaviour reflects `_.uniqBy`:

```
> const { uniqBy } = require('lodash')
> uniqBy( [ null, null, undefined, undefined ] )
[ null, undefined ]
```

File: packages/js-utils/CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+## 1.0.0-alpha.0
```
Comment (scinos) [inline packages/js-utils/CHANGELOG.md]: /nit: is not very common to find packages with `-alpha` or similar. Something like `0.1.0` or `1.0.0` feels more natural to me.

**Review by scinos — APPROVED** [review]

File: packages/js-utils/src/unique-by.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const defaultCompare = ( a: any, b: any ): boolean => a === b;
+
+function uniqueBy( arr: any[], fn = defaultCompare ) {
+	return arr.reduce( ( acc, v ) => {
+		if ( ! acc.some( ( x: any ) => fn( v, x ) ) ) acc.push( v );
+		return acc;
+	}, [] );
+}
```
Comment (sarayourfriend) [inline packages/js-utils/src/unique-by.ts]: I think we can do much better as far as types are concerned here.

```ts
function uniqueBy<T>( arr: T[], fn: ( left: T, right: T ) => boolean ): T[]
```

That should do...

```suggestion
function uniqueBy< T >( arr: T[], fn: ( a: T, b: T ) => boolean = defaultCompare ): T[] {
	return arr.reduce( ( acc, v ) => {
		if ( ! acc.some( ( x: any ) => fn( v, x ) ) ) acc.push( v );
		return acc;
	}, [] as T[] );
}
```

File: packages/js-utils/package.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
+{
+	"name": "@automattic/js-utils",
+	"version": "1.0.0-alpha.0",
+	"description": "General purpose Javascript utility functions",
+	"homepage": "https://github.com/Automattic/wp-calypso",
+	"license": "GPL-2.0-or-later",
+	"author": "Automattic Inc.",
+	"sideEffects": true,
```
Comment (tyxla) [inline packages/js-utils/package.json]: I think ideally this package shouldn't have any side effects. Are we planning any?

File: packages/js-utils/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
+{
+	"compilerOptions": {
+		"target": "ES5",
+		"lib": [ "DOM", "DOM.Iterable", "ESNext" ],
```
Comment (tyxla) [inline packages/js-utils/tsconfig.json]: Do we need those `DOM` libs for anything?

File: packages/js-utils/src/unique-by.ts (hunk 3 above)
Comment (flootr) [reply packages/js-utils/src/unique-by.ts]: Yes, agreed. Although your suggestion shows me there's a syntax error:

Screenshot 2021-02-22 at 15 35 37

should it be this?

```ts
function uniqueBy< T >( arr: T[], fn: ( a: T, b: T ) => boolean = defaultCompare ): T[] {
  // ...
}
```

**Review by tyxla — APPROVED** [review]
Looks good to me!

File: packages/js-utils/tsconfig.json (hunk 5 above)
Comment (flootr) [reply packages/js-utils/tsconfig.json]: I don't think so. Fixed in https://github.com/Automattic/wp-calypso/pull/50263/commits/31b8b3d2521bf99a7a3cb15677166f82c1b65317

File: packages/js-utils/package.json (hunk 4 above)
Comment (flootr) [reply packages/js-utils/package.json]: > Are we planning any?

Nope 👍🏻  fixed in https://github.com/Automattic/wp-calypso/pull/50263/commits/7a468fb549d9ee37b63c278a591880f9755480f4

File: packages/js-utils/src/unique-by.ts (hunk 3 above)
Comment (sarayourfriend) [reply packages/js-utils/src/unique-by.ts]: Yes, sorry!

File: packages/js-utils/CHANGELOG.md (hunk 2 above)
Comment (flootr) [reply packages/js-utils/CHANGELOG.md]: fixed in https://github.com/Automattic/wp-calypso/pull/50263/commits/b9db7131f5b71d1159efdb1f06ab7fd89eda46f2, I don't think there's a need to maintain an alpha version

**Comment by ockham** [comment]
How's ours different from lodash's? https://lodash.com/docs/4.17.15#uniqBy :grimacing:

**Comment by flootr** [comment]
> How's ours different from lodash's? https://lodash.com/docs/4.17.15#uniqBy :grimacing: 

It serves the same purpose. There's an ongoing initiative (mostly lead by @tyxla) to replace some of lodash with either native JS code (which is probably most of the usages) and/or introduce similar functions where this isn't possible. This PR is mostly motivated by not introducing new usages of lodash for some of my other work where I need this functionality.

**Comment by tyxla** [comment]
> > How's ours different from lodash's? https://lodash.com/docs/4.17.15#uniqBy 😬

Ours is essentially much simpler 😉  The Lodash one uses `baseUniq` and that's [far more complex](https://github.com/lodash/lodash/blob/4.17.15/lodash.js#L4249) than we'll ever need here. As you can see, our implementation is much much simpler. 

I'm also planning to post about some of the motivations here, so stay tuned @ockham 😉

**Comment by sgomes** [comment]
For added motivation on rolling out our own, `lodash`'s `uniqBy` loads up to 12KB (4.6 gzipped) of code: https://bundlephobia.com/result?p=lodash.uniqby@4.7.0

Of course, a lot of the code it pulls in is is already indirectly being used elsewhere in Calypso, but if the goal is to get rid of `lodash` altogether, we should definitely target functions like `uniqBy` first, which end up pulling in a disproportionately large chunk of the library.

File: packages/js-utils/src/unique-by.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+const defaultCompare = ( a: any, b: any ): boolean => a === b;
+
+function uniqueBy< T >( arr: T[], fn: ( a: T, b: T ) => boolean = defaultCompare ): T[] {
+	return arr.reduce( ( acc, v ) => {
+		if ( ! acc.some( ( x: any ) => fn( v, x ) ) ) acc.push( v );
```
Comment (nb) [inline packages/js-utils/src/unique-by.ts:5]: This looks like `O(n²)` in all cases. Via an `Object`/`Set` maybe we can make this to always be `O(n)`?

Comment (nb) [reply packages/js-utils/src/unique-by.ts:5]: A note that in this case we would need to move to a key function vs. the compare one, but given that we wanted to mimic `uniqBy` and not `uniqWith` that seems sensible trade-off.

File: packages/js-utils/src/unique-by.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const defaultCompare = ( a: any, b: any ): boolean => a === b;
```
Comment (noahtallen) [inline packages/js-utils/src/unique-by.ts:1]: Should we use `unknown` instead of `any` in this file?

Comment (sarayourfriend) [reply packages/js-utils/src/unique-by.ts:1]: Are you allowed to do comparison operations on `unknown`?

Update: you are! https://www.typescriptlang.org/play?#code/MYewdgzgLgBAJgUwGYEMCuAbKBhEBbABxQCcEYBeGAChQC4Y0wBrMEAdzABoYAjexluzABKCgD4YKCuUo8A3ACggA

I didn't know that, `unknown` would definitely be preferred in that case.

Comment (jsnajdr) [reply packages/js-utils/src/unique-by.ts:1]: Another option, probably the cleanest type-wise, is to make the function generic:
```ts
function defaultComp< A, B >( a: A, b: B ) {
  return a === b;
}
```

Comment (flootr) [reply packages/js-utils/src/unique-by.ts:1]: Oops, I totally missed that feedback. I'll make sure to follow up with a PR!

File: packages/js-utils/src/unique-by.ts (hunk 6 above)
Comment (flootr) [reply packages/js-utils/src/unique-by.ts:5]: I'm sorry I totally missed that feedback. I'll make sure to have a look at it 👍🏻

Comment (nb) [reply packages/js-utils/src/unique-by.ts:5]: No worries!

**Pull request merged**